crates/ruff_python_semantic/src/definition.rs

```diff
@@ -0,0 +1,213 @@
+        &self.0[usize::from(index)]
+    }
+}
+
+impl<'a> Deref for Definitions<'a> {
+    type Target = [Definition<'a>];
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+pub struct DefinitionsIter<'a> {
+    inner: std::slice::Iter<'a, Definition<'a>>,
+    definitions: Vec<&'a Definition<'a>>,
+    visibilities: Vec<Visibility>,
+}
+
+impl<'a> Iterator for DefinitionsIter<'a> {
+    type Item = (&'a Definition<'a>, Visibility);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let definition = self.inner.next()?;
+
+        // Determine the visibility of the next definition, taking into account its parent's
+        // visibility.
+        let visibility = {
+            match definition {
+                Definition::Module(module) => module_visibility(&module.source),
+                Definition::Member(member) => match member.kind {
+                    MemberKind::Class => {
+                        if self.visibilities[usize::from(member.parent)].is_private() {
+                            Visibility::Private
+                        } else {
+                            class_visibility(member.stmt)
+                        }
+                    }
+                    MemberKind::NestedClass => {
+                        if self.visibilities[usize::from(member.parent)].is_private()
+                            || matches!(
+                                self.definitions[usize::from(member.parent)],
+                                Definition::Member(Member {
+                                    kind: MemberKind::Function
+                                        | MemberKind::NestedFunction
+                                        | MemberKind::Method,
+                                    ..
+                                })
+                            )
+                        {
+                            Visibility::Private
+                        } else {
+                            class_visibility(member.stmt)
+                        }
+                    }
+                    MemberKind::Function => {
+                        if self.visibilities[usize::from(member.parent)].is_private() {
+                            Visibility::Private
+                        } else {
+                            function_visibility(member.stmt)
+                        }
+                    }
+                    MemberKind::NestedFunction => Visibility::Private,
+                    MemberKind::Method => {
+                        if self.visibilities[usize::from(member.parent)].is_private() {
+                            Visibility::Private
+                        } else {
+                            method_visibility(member.stmt)
+                        }
+                    }
+                },
+            }
+        };
+        self.definitions.push(definition);
+        self.visibilities.push(visibility);
+
+        Some((definition, visibility))
+    }
```

Comment:
Here, we iterate over definitions and compute visibilities as we go. We can do it in a single pass, as it's assumed that parents appear before children (this assumption is documented above, on `Definitions`).